explorer/interpreter/value.h

```diff
@@ -1188,21 +1188,31 @@ class StaticArrayType : public Value {
  public:
   // Constructs a statically-sized array type with the given element type and
   // size.
-  StaticArrayType(Nonnull<const Value*> element_type, size_t size)
+  StaticArrayType(Nonnull<const Value*> element_type, size_t size,
+                  bool implicit)
       : Value(Kind::StaticArrayType),
         element_type_(element_type),
-        size_(size) {}
+        size_(size),
+        implicit_(implicit) {}
 
   static auto classof(const Value* value) -> bool {
     return value->kind() == Kind::StaticArrayType;
   }
 
   auto element_type() const -> const Value& { return *element_type_; }
+  auto set_size(size_t size) const -> void {
+    const_cast<StaticArrayType*>(this)->size_ = size;
+  }
```

Comment:
This operation and `set_explicit` make me nervous. They're marked `const`, but it sure seems like they're mutating the object. Why exactly do we need them?

My hunch is that this should really be handled as part of type deduction, which already supports filling in unknown types without mutating them.

Reply:
Can you point out how Carbon handles type deduction? I have been looking around the concept of type deduction in `type_checker.cpp` but got no result.

Reply:
See `TypeChecker::ArgumentDeduction`.

Reply:
I have been looking around the actual mutation of type deduction in `TypeChecker::ArgumentDeduction`, I still failed to find a proper way to perform deduction, can you show me more specific way to do so? I would appreciate that.

Reply:
It might be helpful to follow how `auto` gets its type assigned, since that's equivalent behavior (receiving its type from the rhs of `=`).

Also, regarding `Value` mutability -- I think @geoffromer is right to be nervous here, `Value` objects should represent a singular value, i.e. constant. In the interpreter, mutations occur by creating new `Value` objects.